### PR TITLE
Object Oriented Refactor

### PR DESCRIPTION
**Pull request by AR2100**
# Description

This PR creates an object-oriented structure to the codebase. Code is organized into 2 primary folders: drivers and managers. The drivers interact with the low-level hardware, such are the Ethernet interface and UART. The drivers are instantiated in `Context.hpp`. The managers use the drivers to get useful work done. They are instantiated in `main.cpp` and access the drivers through the context object. 

This PR does the following:
- Splits existing EthernetDriver code into driver and manager paradigm
- Adds Roboclaw support for speed control

# Self Checklist
- [x] I have tested that the new behavior works
- [x] I have formatted the code with `make format`

### CODE REVIEW

File: include/main.hpp [hunk 1]
```diff
@@ -1,14 +1,23 @@
...
-#include "urc.pb.h"
+#include "Messages.hpp"
+#include "Managers/Managers.hpp"
 
+struct RoboClawData {
+  uint8_t wheelSpeed;
+  bool valid;
+};
+
+void setupRoboClaw(Context &context);
```
Comment (a-stickan) [inline include/main.hpp]: What are all these methods being declared for? They don't appear in the actual .cpp file

File: include/Managers/Managers.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    const static size_t MAX_NAME_LEN = 30;
+    motors::MotorController *_motorController;
+    uint32_t _address;
+    uint32_t _channel;
+    char _name[MAX_NAME_LEN];
+  };
+  Motor _motors[NUM_MOTORS];
+};
+
+class EthernetManager {
```
Comment (a-stickan) [inline include/Managers/Managers.hpp]: You sure this class should be part of this file? I'd go over your header files, some of them seem a little general for my taste. I always go with the general rule of a header file for every class for simplicity

File: include/Managers/Managers.hpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+#define MANAGERS_HPP
+
+#include <Arduino.h>
+#include "Context.hpp"
+#include "Messages.hpp"
+#include "Drivers/MotorController.hpp"
+
+namespace manager {
+
+const size_t NUM_MOTORS = 2;
```
Comment (a-stickan) [inline include/Managers/Managers.hpp]: Should this be part of the motor manager class? Don't think it should belong to the manager namespace in general

File: include/Drivers/MotorController.hpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
+#ifndef MOTORCONTROLLER_HPP
+#define MOTORCONTROLLER_HPP
+
+#include <Arduino.h>
+#include "RoboClaw.h"
+
+namespace motors {
+
+enum MotorMode { TORQUE, SPEED, VELOCITY };
```
Comment (a-stickan) [inline include/Drivers/MotorController.hpp]: I can't find any place where this is being used. Why is it here?

File: include/Drivers/MotorController.hpp [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  MotorController(){};
+  virtual void begin(long speed);
+  // TODO: virtual void setMode(uint8_t channel, MotorMode mode) = 0;
+  virtual int32_t getSpeed(uint8_t address, uint8_t channel, bool &valid) = 0;
+  virtual void setSpeed(int32_t speed, uint8_t address, uint8_t channel, bool &valid) = 0;
+  // TODO: virtual int32_t getPosition(uint8_t channel) = 0;
+  // TODO: virtual void setPosition(uint8_t channel) = 0;
+};
+
+class RoboClawController : public MotorController {
```
Comment (a-stickan) [inline include/Drivers/MotorController.hpp]: Once again I'd probably separate these classes into different files. It makes your architecture a little hard to follow unless you know it well

**Review by a-stickan — CHANGES_REQUESTED** [review]
Header files are a little confusing at the moment. Also maybe look into your comments, they were a little excessive in some of the files

File: include/main.hpp (hunk 1 above)
Comment (AR2100) [reply include/main.hpp]: Not necessary, deleted

Comment (AR2100) [reply include/main.hpp:16]: Deleted

File: include/Drivers/MotorController.hpp (hunk 4 above)
Comment (AR2100) [reply include/Drivers/MotorController.hpp]: Not used, just deleted

File: include/Managers/Managers.hpp (hunk 3 above)
Comment (AR2100) [reply include/Managers/Managers.hpp]: You're right, made it a member of motor manager.

**Review by AR2100 — COMMENTED** [review]
What do you think of the file structure now @a-stickan ?

I removed all of the excessive comments from the header files, was messing around with Doxygen.

**Review by a-stickan — APPROVED** [review]

**Comment by a-stickan** [comment]
Everything looks good to merge!
